### DISCUSSION TITLE
Intermittence SocketException: Connection reset by peer

### DISCUSSION THREAD

**Discussion by codeeuser** [serverpod/serverpod#3028] [Q&A] [answered]
I have encountered this issue while I deployed the Serverpod on Linux Server

SERVERPOD VERSION: 2.2.0

ERROR:
SocketException: Connection reset by peer (OS Error: Connection reset by peer, errno = 104)

SUGGESTION:
Upgrade the code on the file below:
path: serverpod-2.2.0/packages/serverpod/lib/src/server/server.dart

    while (true) { // <==add this infinite loop in case the exception and restart the server
      try {
        await for (var request in httpServer) {
          serverpod.logVerbose(
            'received request: ${request.method} ${request.uri.path}',
          );

          _handleRequestWithErrorBoundary(request);
        }
      } catch (e, stackTrace) {
        stderr.writeln(
            '${DateTime.now().toUtc()} Internal server error. httpSever.listen failed.');
        stderr.writeln('$e');
        stderr.writeln('$stackTrace');
        await shutdown(); // <==add this code
        await start(); // <==add this code
      }
    }
  }

**Comment by SandPod** [comment]
Hey @codeeuser!

Additional questions:
- Is there a stack trace with the error that you could supply?
- Is there any pattern on when this happens? When you start the server or when the server has been running for a while?
- Could this connection issue stem from the connection to the database? If so, could this issue potentially be related? https://github.com/serverpod/serverpod/issues/2965

**Comment by codeeuser** [comment]
SERVER CPU: 1CPU
SERVER Memory: 1G

Is there a stack trace with the error that you could supply?
"SocketException: Connection reset by peer (OS Error: Connection reset by peer, errno = 104), address = ::, port = 7081"
I searched the keyword from the serverpod version 2.2.0 source code and found the only location on server.dart

Is there any pattern on when this happens? When you start the server or when the server has been running for a while?
I could not found the pattern. The error came out in uncertain duration. Only I suspected was the timeout on somewhere else, or may be server memory was not enough. I also checked the Linux syslog and configured verbose logging. If memory is not enough, is there any best garbage collection system can be implemented? 

Could this connection issue stem from the connection to the database? If so, could this issue potentially be related?
I read the issue from the Github and updated the localhost to 127.0.0.1, but still got this error.

**Comment by codeeuser** [comment]
UPDATE:
I have update my server code for reusing the Stream (WebSocket) instance for prevent the memory issue. I hope I need few days for verify it and waiting my issue solved. However, I suggest Serverpod can handle the failure of httpserver instance as well.

**Comment by SandPod** [comment]
Thanks for the additional information!

### Additional questions

How is your load balancer configured? A common culprit when working with websockets is that the load balancer can be configured to automatically close websockets after a certain amount of time. I'm not entirely convinced that would manifest in this way but it could be worth checking.

When the issue happens, does the server completely shut down, or is it placed in a state that it is not possible to recover from? 

### Related issues

~~There is an issue reported with the dart SDK for the same issue~~ There are multiple issues open in the Dart SDK for similar issues:
- https://github.com/dart-lang/sdk/issues/55776
- https://github.com/dart-lang/sdk/issues/47247
- https://github.com/dart-lang/sdk/issues/31558
- https://github.com/dart-lang/sdk/issues/27440

There was also recently a memory leak issue that was resolved in dart around sockets that could probably lead to unexpected behavior: https://github.com/serverpod/serverpod/issues/762.

**Comment by codeeuser** [comment]
I am investigating my code has not canceled the stream connection. I need time for verify it.

**Comment by codeeuser** [comment]
This issue was resolved.

SOLUTION:
Cancel the stream on the dispose method on Flutter client side instead of listening the stream with infinite 'FOR' loop.

**Comment by SandPod** [comment]
Oh, good that it was resolved!

Are you using the old streaming interface where you manage the connection manually?

**Comment by codeeuser** [comment, marked as answer]
The Stream must be canceled while the widget dispose, but not waiting the server side timeout.

LATEST CODE:

StreamSubscription<CountToken>? _streamSubscription;

@override
  void dispose() async {
    await _streamSubscription?.cancel(); // <== MUST CANCEL the STREAM
    super.dispose();
  }

Future<void> _listenToUpdates(int windowId) async {
    final Stream<CountToken> statusUpdates = client.tokenIssued.echoStatusStream();
    try {
      _streamSubscription = statusUpdates.listen((update) {
        // GET THE UPDATE
      });
    } catch (e) {
      // ERROR
    }
  }

**Comment by SandPod** [comment]
That sounds reasonable, will convert this issue to a Q&A and mark your reply as the answer.
